datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -2110,16 +2110,38 @@ pub fn cardinality(args: &[ArrayRef]) -> Result<ArrayRef> {
         return exec_err!("cardinality expects one argument");
     }
 
-    let list_array = as_list_array(&args[0])?.clone();
-
-    let result = list_array
-        .iter()
-        .map(|arr| match compute_array_dims(arr)? {
-            Some(vector) => Ok(Some(vector.iter().map(|x| x.unwrap()).product::<u64>())),
-            None => Ok(None),
-        })
-        .collect::<Result<UInt64Array>>()?;
-
+    let result = match &args[0].data_type() {
+        DataType::List(_) => {
```

Comment:
Although this implementation is small, I think having a closure for generic array is still a good practice.

Reply:
I agree it would be better to avoid the duplication if possible